src/beamlime/applications/_nexus_helpers.py

```diff
@@ -0,0 +1,303 @@
+    cur_idx = _cur_idx or tuple()
+
+    if not filter_or_indices:
+        return cur_idx
+
+    cur_filter, _left_filters = filter_or_indices[0], filter_or_indices[1:]
+    if (isinstance(obj, dict) and isinstance(cur_filter, str)) or (
+        isinstance(obj, list) and isinstance(cur_filter, int)
+    ):
+        return find_index(
+            obj[cur_filter], *_left_filters, _cur_idx=cur_idx + (cur_filter,)
+        )
+    elif isinstance(obj, dict) and callable(cur_filter):
+        key, value = next((k, v) for k, v in obj.items() if cur_filter(k, v))
+        return find_index(value, *_left_filters, _cur_idx=cur_idx + (key,))
+    elif isinstance(obj, list) and callable(cur_filter):
+        idx, value = next(((i, v) for i, v in enumerate(obj) if cur_filter(v)))
+        return find_index(value, *_left_filters, _cur_idx=cur_idx + (idx,))
+    else:
+        raise KeyError(
+            f"Invalid filter function or index: {cur_filter} for type: {type(obj)}"
+        )
+
+
+def nested_dict_getitem(obj: dict, *indexes: str | int) -> Any:
+    """Get the item from the nested dictionary."""
+    if not indexes:
+        return obj
+    return nested_dict_getitem(obj[indexes[0]], *indexes[1:])
+
+
+def nested_dict_select(
+    obj: dict, *filter_or_indices: Callable[..., bool] | str | int
+) -> Any:
+    """Select the first matching item from the nested dictionary."""
+    return nested_dict_getitem(obj, *find_index(obj, *filter_or_indices))
+
+
+def match_nx_class(child: dict, cls_name: str) -> bool:
+    attrs = child.get('attributes', [])
+    return any(
+        attr.get('name', None) == "NX_class" and attr.get("values", None) == cls_name
+        for attr in attrs
+    )
+
+
+def _match_dataset_name(child: dict, name: str) -> bool:
+    return (
+        child.get('module', None) == 'dataset'
+        and child.get('config', {}).get('name', None) == name
+    )
+
+
+def _match_module_name(child: dict, name: str) -> bool:
+    return child.get('module', None) == name
+
+
+def _match_dataset_name_suffix(child: dict, suffix: str) -> bool:
+    return child.get('module', None) == 'dataset' and child.get('config', {}).get(
+        'name', ''
+    ).endswith(suffix)
+
+
+def collect_nested_module_indices(
+    nexus_container: dict, fb_id: FBIdentifier
+) -> list[tuple]:
+    """Collect the indices of the module with the given id."""
+    collected = list()
+
+    def _recursive_collect_indices(
+        obj: dict,
+        cur_idx: tuple[str | int, ...],
+    ) -> None:
+        # Group
+        for i, cand in enumerate(obj.get('children', [])):
+            cand_idx = cur_idx + ('children', i)
+            if _match_module_name(cand, fb_id):
+                collected.append(cand_idx)
+            else:
+                _recursive_collect_indices(cand, cand_idx)
+
+    _recursive_collect_indices(nexus_container, tuple())
+    return collected
+
+
+class ModularDatasetContainer(dict):
```

Comment:
Where is this used?

Reply:
Sorry these are for the #134 I'll move it